### PR TITLE
Add AttributeMap and use it in the Resource

### PR DESCRIPTION
**Pull request by bogdandrutu**
This PR plugs the new AttributeMap only into the Resource, but we can see a very nice benefit from this in the metrics benchmarks:

Before:
```
go test -benchmem -bench=. -run=notests ./internal/data/...
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector/internal/data
BenchmarkOtlpToFromInternal_PassThrough-16                              318627304                3.66 ns/op            0 B/op          0 allocs/op
BenchmarkOtlpToFromInternal_Int64Points_MutateOneLabel-16                2147724               551 ns/op             688 B/op         13 allocs/op
BenchmarkOtlpToFromInternal_DoublePoints_MutateOneLabel-16               2170123               554 ns/op             688 B/op         13 allocs/op
BenchmarkOtlpToFromInternal_HistogramPoints_MutateOneLabel-16            1902834               638 ns/op             768 B/op         14 allocs/op
BenchmarkOtlpToFromInternal_SummaryPoints_MutateOneLabel-16              1880725               639 ns/op             768 B/op         14 allocs/op
PASS
ok      github.com/open-telemetry/opentelemetry-collector/internal/data 11.143s
```

After:
```
go test -benchmem -bench=. -run=notests ./internal/data/...
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector/internal/data
BenchmarkOtlpToFromInternal_PassThrough-16                              1000000000               0.256 ns/op           0 B/op          0 allocs/op
BenchmarkOtlpToFromInternal_Int64Points_MutateOneLabel-16                4681989               252 ns/op             240 B/op          7 allocs/op
BenchmarkOtlpToFromInternal_DoublePoints_MutateOneLabel-16               4736796               249 ns/op             240 B/op          7 allocs/op
BenchmarkOtlpToFromInternal_HistogramPoints_MutateOneLabel-16            3651998               331 ns/op             320 B/op          8 allocs/op
BenchmarkOtlpToFromInternal_SummaryPoints_MutateOneLabel-16              3555200               336 ns/op             320 B/op          8 allocs/op
PASS
ok      github.com/open-telemetry/opentelemetry-collector/internal/data 7.523s
```

### CODE REVIEW

File: internal/data/common.go [hunk 1]
```diff
@@ -126,7 +128,161 @@ func (a AttributeValue) SetBool(v bool) {
...
+func (akv AttributeKeyValue) SetValue(av AttributeValue) {
+	akv.orig.Type = av.orig.Type
+	akv.orig.StringValue = av.orig.StringValue
+	akv.orig.IntValue = av.orig.IntValue
+	akv.orig.DoubleValue = av.orig.DoubleValue
+	akv.orig.BoolValue = av.orig.BoolValue
+}
+
+// AttributeMap stores a map of attribute keys to values.
+type AttributeMap struct {
```
Comment (dmitryax) [inline internal/data/common.go]: The name looks a bit confusing. I see it does implement an interface similar to map, but did you consider any other name options?

Comment (bogdandrutu) [reply internal/data/common.go]: I did not consider other options, I like the current name (personal opinion). Do you have any suggestions (open to change if you have a better suggestion)?

**Comment by bogdandrutu** [comment]
Based on https://github.com/open-telemetry/opentelemetry-collector/issues/653 I propose that:
1. First options is to have all the methods in the map to accept the immutable types
`InsertInt()`
`InsertBool()`
`InsertDouble()`
`InsertString()`
1. Second option is to make AttributeValue a real struct not backed by the proto (avoid unnecessary allocation that we cannot use it anyway). Then we can keep all the methods as they are.

File: internal/data/common.go (hunk 1 above)
Comment (bogdandrutu) [reply internal/data/common.go]: `AttributeSliceMap`? :)

Comment (tigrannajaryan) [reply internal/data/common.go]: I'd suggest to keep the name. The fact that it uses a slice is an implementation detail and may even change in the future.
Semantics and the set of operations this type implements are of a classic Map.

File: internal/data/common.go [hunk 2]
```diff
@@ -126,7 +128,161 @@ func (a AttributeValue) SetBool(v bool) {
...
+// NewAttributeMap creates a new AttributeMap from the given map[string]AttributeValue.
+func NewAttributeMap(attrMap map[string]AttributeValue) AttributeMap {
+	origs := make([]otlpcommon.AttributeKeyValue, len(attrMap))
+	wrappers := make([]*otlpcommon.AttributeKeyValue, len(attrMap))
+
+	ix := 0
+	for k, v := range attrMap {
+		wrappers[ix] = &origs[ix]
+		wrappers[ix].Key = k
+		AttributeKeyValue{wrappers[ix]}.SetValue(v)
```
Comment (tigrannajaryan) [inline internal/data/common.go]: Perhaps simpler to set Value directly.

File: internal/data/common.go (hunk 1 above)
Comment (dmitryax) [reply internal/data/common.go]: I was thinking of something like just `Attributes`, but it's good enough. Let's keep `AttributesMap`

**Comment by bogdandrutu** [comment]
@tigrannajaryan any comment about https://github.com/open-telemetry/opentelemetry-collector/pull/650#issuecomment-600811329?

File: internal/data/common.go [hunk 3]
```diff
@@ -126,7 +128,161 @@ func (a AttributeValue) SetBool(v bool) {
...
+	return AttributeKeyValue{nil}, false
+}
+
+// Delete deletes the entry associated with the key and returns true if the key
+// was present in the map, otherwise returns false.
+func (am AttributeMap) Delete(key string) bool {
+	for i, a := range *am.orig {
+		if a.Key == key {
+			(*am.orig)[i] = (*am.orig)[len(*am.orig)-1]
+			*am.orig = (*am.orig)[:len(*am.orig)-1]
```
Comment (pjanotti) [inline internal/data/common.go]: nit: this shouldn't make any practical difference due to the way that they are allocated but in the unlike case that the pointer put in index `i` is later released adding `(*am.orig)[len(*am.orig)-1] = nil` ensures no lingering references from the part of the slice that is not used anymore but still on memory.

**Pull request merged**